url.go

```diff
@@ -0,0 +1,309 @@
+package imgproxy
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// ImgproxyURLData is a struct that contains the data required for generating an imgproxy URL.
+type ImgproxyURLData struct {
+	*Imgproxy
+	Options map[string]string
+}
+
+const insecureSignature = "insecure"
+
+// Generate generates the imgproxy URL.
+func (i *ImgproxyURLData) Generate(uri string) (string, error) {
+	if i.cfg.Encode {
+		uri = base64.RawStdEncoding.EncodeToString([]byte(uri))
+	} else {
+		uri = "plain/" + uri
+	}
+
+	keys := make([]string, len(i.Options))
+	j := 0
+	for key := range i.Options {
+		keys[j] = key
+		j++
+	}
+	sort.Strings(keys)
+
+	options := "/"
+	for _, key := range keys {
+		options += key + ":" + i.Options[key] + "/"
+	}
+
+	uriWithOptions := options + uri
+
+	if len(i.salt) == 0 && len(i.key) == 0 {
+		return i.cfg.BaseURL + insecureSignature + uriWithOptions, nil
+	}
+
+	signature, err := getSignatureHash(i.key, i.salt, i.cfg.SignatureSize, uriWithOptions)
+	if err != nil {
+		return "", err
+	}
+
+	return i.cfg.BaseURL + signature + uriWithOptions, nil
+}
+
+func getSignatureHash(key []byte, salt []byte, signatureSize int, payload string) (string, error) {
+	signature := hmac.New(sha256.New, key)
+
+	if _, err := signature.Write(salt); err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	if _, err := signature.Write([]byte(payload)); err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	sha := base64.RawURLEncoding.EncodeToString(signature.Sum(nil)[:signatureSize])
+
+	return sha, nil
+}
+
+// ResizingType enum.
+type ResizingType string
+
+// ResizingType enum.
+const (
+	ResizingTypeFit      = ResizingType("fit")       // resizes the image while keeping aspect ratio to fit a given size.
+	ResizingTypeFill     = ResizingType("fill")      // resizes the image while keeping aspect ratio to fill a given size and crops projecting parts.
+	ResizingTypeFillDown = ResizingType("fill-down") // the same as fill, but if the resized image is smaller than the requested size, imgproxy will crop the result to keep the requested aspect ratio.
+	ResizingTypeForce    = ResizingType("force")     // resizes the image without keeping the aspect ratio.
+	ResizingTypeAuto     = ResizingType("auto")      // if both source and resulting dimensions have the same orientation (portrait or landscape), imgproxy will use fill. Otherwise, it will use fit.
```

Comment:
Nit.

My IDE (VSCode) doesn't show the comment when hovering over these constants elsewhere in the code. If you put the comment above it (on a separate line), it will work:

```suggestion
	// Resizes the image while keeping aspect ratio to fit a given size.
	ResizingTypeFit = ResizingType("fit")
	// Resizes the image while keeping aspect ratio to fill a given size and crops projecting parts.
	ResizingTypeFill = ResizingType("fill")
	// The same as fill, but if the resized image is smaller than the requested size, imgproxy will crop the result to keep the requested aspect ratio.
	ResizingTypeFillDown = ResizingType("fill-down")
	// Resizes the image without keeping the aspect ratio.
	ResizingTypeForce = ResizingType("force")
	// If both source and resulting dimensions have the same orientation (portrait or landscape), imgproxy will use fill. Otherwise, it will use fit.
	ResizingTypeAuto = ResizingType("auto")
```